### ISSUE TITLE
Type guard regression on 1.1.159

### ISSUE THREAD

**Issue by bellini666** [microsoft/pyright#2137]
Not sure how to properly describe the error itself, but I just notice this regression:

```python
from typing import Optional

d: dict[str, int] = {}
mapper: dict[str, Optional[str]] = {}

def foo(bar: list[str]):
    return sum(d[mapper[i]] for i in bar if mapper[i] is not None)
```

When running pyright I get:

```
❯ pyright foo.py
No configuration file found.
No pyproject.toml file found.
stubPath /tmp/xxx/typings is not a valid directory.
Assuming Python platform Linux
Searching for source files
Found 1 source file
/tmp/xxx/foo.py
  /tmp/xxx/foo.py:7:16 - error: Argument of type "str | None" cannot be assigned to parameter "k" of type "str" in function "__getitem__"
    Type "str | None" cannot be assigned to type "str"
      Type "None" cannot be assigned to type "str" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos 
Completed in 0.653sec
```

Which is not considering the `if mapper[i] is not None` which makes it for sure a `str` and not `None`

This was not an error on 1.1.158

**Comment by erictraut** [comment]
This should not have worked in previous versions. If it did, that was a bug.

You are using a subscript expression `mapper[i]` with a dynamic index. Pyright performs assignment type narrowing for subscript expressions only when the subscript is an integer or string literal (like `mapper[1]`). When the subscript is dynamic, it's unsafe to perform narrowing.

If you want to use the above pattern, the recommended workaround is to use the walrus operator, like this:

```python
def foo(bar: list[str]):
    return sum(d[x] for i in bar if (x := mapper[i]) is not None)
```

Not only will this address the type issue, it will also make your code run faster.

**Comment by bellini666** [comment]
Ok, I got your point.

And thanks for the tip. I'm an old school python developer, still didn't get used to that `x:= y` expression assignment thing yet. Interesting way to use that there
